### PR TITLE
Re-Implementation of reverse string via CFString

### LINKED ISSUE #2478: NSString _reverseString should be implemented via CFString

### PR DESCRIPTION
**Pull request by msft-Jeyaram**
- Implement _reverseString via CoreFoundation and _wcsrev
- Remove _reverseString from Foundation (only CoreText uses it)
- CTline _getTruncatedStringFromSourceLine should use CF*based reverse string
- Remove redundant tests

fixes #2478

### CODE REVIEW

**Comment by DHowett-MSFT** [comment]
nit: reword to be imperative mood

**Comment by msft-Jeyaram** [comment]
Getting some bench mark for old vs new

File: Frameworks/CoreFoundationAdditions/CFStringHelpers.mm [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+CFStringRef _CFStringCreateReversedString(CFStringRef string) {
+    if (string == nullptr) {
+        return nullptr;
+    }
+
+    CFIndex index = CFStringGetLength(string);
+
+    if (index < 2) {
+        return CFStringCreateCopy(kCFAllocatorDefault, string);
```
Comment (DHowett-MSFT) [inline Frameworks/CoreFoundationAdditions/CFStringHelpers.mm]: this could also be CFRetain

File: Frameworks/CoreFoundationAdditions/CFStringHelpers.mm [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    }
+
+    CFMutableStringRef result = CFStringCreateMutable(kCFAllocatorDefault, 0);
```
Comment (DHowett-MSFT) [inline Frameworks/CoreFoundationAdditions/CFStringHelpers.mm]: If you create the new string with a capacity hint from the original string, you can avoid the inevitable reallocations

File: Frameworks/CoreFoundationAdditions/CFStringHelpers.mm [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+    while (index > 0) {
```
Comment (DHowett-MSFT) [inline Frameworks/CoreFoundationAdditions/CFStringHelpers.mm]: it may be cheaper to create a unichar buffer and pass that off to CFStringCreateWithCharacters -- you get the added benefit of not creating a _mutable_ string

It should be somewhat "safe" to use utf-16 codepoints or unichars; it's definitely safer than utf-8

Comment (rajsesh-msft) [inline Frameworks/CoreFoundationAdditions/CFStringHelpers.mm]: This is going to suck when it comes to perf.

**Comment by DHowett-MSFT** [comment]
Also, I wonder how it works with Composing Diacritics! It might be reasonable to normalize the string first...

File: Frameworks/CoreFoundationAdditions/CFStringHelpers.mm [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+// THE SOFTWARE.
+//
+//******************************************************************************
+
+#import <CoreFoundation/CFString.h>
+#import <Starboard/SmartTypes.h>
+#import <CFInternal.h>
+#import "CFFoundationInternal.h"
+
+CFStringRef _CFStringCreateReversedString(CFStringRef string) {
```
Comment (rajsesh-msft) [inline Frameworks/CoreFoundationAdditions/CFStringHelpers.mm]: Given that there is exactly one place where this is used (and was used in the past), I don't like that we are doing this.  Why not have the helper in CT and just use it there?

File: Frameworks/CoreFoundationAdditions/CFStringHelpers.mm (hunk 1 above)
Comment (msft-Jeyaram) [inline Frameworks/CoreFoundationAdditions/CFStringHelpers.mm]: If the input is a mutable string, retain might not be a good idea/

Comment (DHowett-MSFT) [inline Frameworks/CoreFoundationAdditions/CFStringHelpers.mm]: Good point!

File: Frameworks/CoreFoundationAdditions/CFStringHelpers.mm (hunk 3 above)
Comment (msft-Jeyaram) [inline Frameworks/CoreFoundationAdditions/CFStringHelpers.mm]: that's what i was going with first, it might be more sensible given that only CT uses it.
Hmm i might as well move it in there, but this kind of makes it more util based.

File: Frameworks/CoreFoundationAdditions/CFStringHelpers.mm (hunk 4 above)
Comment (msft-Jeyaram) [inline Frameworks/CoreFoundationAdditions/CFStringHelpers.mm]: That's what my initial plan was, but since this seems very common and could be utilized by others, it would make a good addition. Also we already have some tests that can be tested independently.

File: Frameworks/CoreFoundationAdditions/CFStringHelpers.mm (hunk 3 above)
Comment (msft-Jeyaram) [inline Frameworks/CoreFoundationAdditions/CFStringHelpers.mm]: YEP utf-16 is safer than utf-8 :)

---
In reply to: [110780284](https://github.com/Microsoft/WinObjC/pull/2479#discussion_r110780284) [](ancestors = 110780284)

Comment (msft-Jeyaram) [inline Frameworks/CoreFoundationAdditions/CFStringHelpers.mm]: might as well use buffer indexing, given it's somewhat "safe"

---
In reply to: [110780304](https://github.com/Microsoft/WinObjC/pull/2479#discussion_r110780304) [](ancestors = 110780304)

File: Frameworks/CoreFoundationAdditions/CFStringHelpers.mm (hunk 4 above)
Comment (rajsesh-msft) [inline Frameworks/CoreFoundationAdditions/CFStringHelpers.mm]: Lets not do this.  the _reverseString was added because some other code was using it.  Given that there is only one instance where _reverseString was used, there is no need to go overboard here.  

---
In reply to: [110781327](https://github.com/Microsoft/WinObjC/pull/2479#discussion_r110781327) [](ancestors = 110781327)

Comment (rajsesh-msft) [inline Frameworks/CoreFoundationAdditions/CFStringHelpers.mm]: also, what is wrong with good old fashioned strrev variants?

---
In reply to: [110782273](https://github.com/Microsoft/WinObjC/pull/2479#discussion_r110782273) [](ancestors = 110782273,110781327)

Comment (msft-Jeyaram) [inline Frameworks/CoreFoundationAdditions/CFStringHelpers.mm]: _strrev + UTF-16 encoding should do the trick then.

---
In reply to: [110782324](https://github.com/Microsoft/WinObjC/pull/2479#discussion_r110782324) [](ancestors = 110782324,110782273,110781327)

**Comment by msft-Jeyaram** [comment]
@rajsesh-msft  still need reverse in CFString backed by strrev, that is contained in CoreText.
Updating the CR

File: Frameworks/CoreText/CTUtilities.mm [hunk 5]
```diff
@@ -15,7 +15,10 @@
 //******************************************************************************
 
 #import <CoreText/CTUtilities.h>
+#import <Starboard/SmartTypes.h>
```
Comment (msft-Jeyaram) [inline Frameworks/CoreText/CTUtilities.mm]: removed

File: Frameworks/CoreText/CTLine.mm [hunk 6]
```diff
@@ -188,8 +187,9 @@ CTLineRef CTLineCreateTruncatedLine(CTLineRef sourceLine, double width, CTLineTr
         CFDictionaryRef attribs = CTRunGetAttributes(currentRun);
 
         if (truncationType == kCTLineTruncationStart) {
-            NSString* reverse = [runString _reverseString];
-            NSAttributedString* string = [[NSAttributedString alloc] initWithString:reverse attributes:(NSDictionary*)attribs];
+            auto reverse = woc::MakeStrongCF<CFStringRef>(_CTCreateReversedString(static_cast<CFStringRef>(runString)));
```
Comment (ms-jihua) [inline Frameworks/CoreText/CTLine.mm]: if there's just the one usage of it, can it be a file-scope function instead?

File: Frameworks/CoreText/CTUtilities.mm [hunk 7]
```diff
@@ -24,3 +27,24 @@
...
 }
+
+// Create a reversed version of the string
+CFStringRef _CTCreateReversedString(CFStringRef string) {
+    if (string == nullptr) {
+        return nullptr;
+    }
+
+    // Returns the number (in terms of UTF-16 code pairs) of Unicode characters in a string.
+    CFIndex length = CFStringGetLength(string);
```
Comment (msft-Jeyaram) [inline Frameworks/CoreText/CTUtilities.mm]: using CFStringGetBytes to be safer

**Comment by msft-Jeyaram** [comment]
@DHowett-MSFT  @ms-jihua  please take a look

File: Frameworks/CoreText/CTLine.mm [hunk 8]
```diff
@@ -16,16 +16,38 @@
...
 #import <StubReturn.h>
-#import "NSStringInternal.h"
 #import "CoreTextInternal.h"
 #import "CGContextInternal.h"
 #import "DWriteWrapper_CoreText.h"
 #import <CoreText/CTTypesetter.h>
 
 #import <algorithm>
 #import <numeric>
+#import <wchar.h>
```
Comment (DHowett-MSFT) [inline Frameworks/CoreText/CTLine.mm]: nit: prefer `<cwchar>`

File: Frameworks/CoreText/CTLine.mm [hunk 9]
```diff
@@ -16,16 +16,38 @@
...
+
+    CFIndex length = CFStringGetLength(string);
+    CFIndex usedBufLen;
+    CFStringGetBytes(string, CFRangeMake(0, length), kCFStringEncodingUTF16, 0, false, nullptr, length, &usedBufLen);
+
+    if (length < 2) {
+        return CFStringCreateCopy(kCFAllocatorDefault, string);
+    }
+
+    std::vector<UniChar> characters((usedBufLen / (sizeof(UniChar))) + 1, '\0');
```
Comment (DHowett-MSFT) [inline Frameworks/CoreText/CTLine.mm]: If you use unique_iw instead of std::vector, you can use `CFStringCreateWithBytesNoCopy` and detach it instead of copying the buffer.

File: Frameworks/CoreText/CTLine.mm [hunk 10]
```diff
@@ -188,8 +210,9 @@ CTLineRef CTLineCreateTruncatedLine(CTLineRef sourceLine, double width, CTLineTr
         CFDictionaryRef attribs = CTRunGetAttributes(currentRun);
 
         if (truncationType == kCTLineTruncationStart) {
-            NSString* reverse = [runString _reverseString];
-            NSAttributedString* string = [[NSAttributedString alloc] initWithString:reverse attributes:(NSDictionary*)attribs];
+            auto reverse = woc::MakeStrongCF<CFStringRef>(_CTCreateReversedString(static_cast<CFStringRef>(runString)));
+            NSAttributedString* string =
+                [[NSAttributedString alloc] initWithString:static_cast<NSString*>(reverse.get()) attributes:(NSDictionary*)attribs];
```
Comment (DHowett-MSFT) [inline Frameworks/CoreText/CTLine.mm]: ??? can't we use CFAttributedStringCreate?

Comment (msft-Jeyaram) [inline Frameworks/CoreText/CTLine.mm]: I already have changes for all NSAttributed, this is strictly for the reverse.

File: Frameworks/CoreText/CTLine.mm [hunk 11]
```diff
@@ -16,16 +16,38 @@
...
 #import <vector>
 
+static CFStringRef _CTCreateReversedString(CFStringRef string) {
+    if (string == nullptr) {
+        return nullptr;
+    }
+
+    CFIndex length = CFStringGetLength(string);
+    CFIndex usedBufLen;
+    CFStringGetBytes(string, CFRangeMake(0, length), kCFStringEncodingUTF16, 0, false, nullptr, length, &usedBufLen);
```
Comment (ms-jihua) [inline Frameworks/CoreText/CTLine.mm]: Seems kind of expensive for this purpose (unless CFStringGetBytes has a fastpath here?)

I wonder if CFStringGetMaximumSizeForEncoding() would be faster/would still work?

Comment (msft-Jeyaram) [inline Frameworks/CoreText/CTLine.mm]: Yes it does take a faster way than if you provide a buffer.
Further improvements can be addressed by a bug to improve core text. The CR is mainly to remove Foundation dependency and what we have now is better than the earlier Foundation/copy based implementation.

File: Frameworks/CoreText/CTLine.mm [hunk 12]
```diff
@@ -16,16 +16,48 @@
...
 #import "CGContextInternal.h"
 #import "DWriteWrapper_CoreText.h"
 #import <CoreText/CTTypesetter.h>
 
 #import <algorithm>
 #import <numeric>
+#import <cwchar>
 #import <vector>
 
+static CFStringRef __CTCreateReversedString(CFStringRef string) {
```
Comment (aballway) [inline Frameworks/CoreText/CTLine.mm]: how is this change being tested?

Comment (msft-Jeyaram) [inline Frameworks/CoreText/CTLine.mm]: UnitTests

---
In reply to: [111007038](https://github.com/Microsoft/WinObjC/pull/2479#discussion_r111007038) [](ancestors = 111007038)

File: Frameworks/CoreText/CTLine.mm [hunk 13]
```diff
@@ -16,16 +16,48 @@
...
+    CFIndex usedBufLen;
+    CFStringGetBytes(string, CFRangeMake(0, length), kCFStringEncodingUTF16, 0, false, nullptr, length, &usedBufLen);
+
+    if (length < 2) {
+        return CFStringCreateCopy(kCFAllocatorDefault, string);
+    }
+
+    CFIndex bufLen = usedBufLen + 1 * sizeof(UniChar);
+    woc::unique_iw<UniChar> characters(static_cast<UniChar*>(IwMalloc(bufLen)));
+    memset(characters.get(), '\0', bufLen);
```
Comment (DHowett-MSFT) [inline Frameworks/CoreText/CTLine.mm]: c'mon, you only need to set the last character in the buffer

File: Frameworks/CoreText/CTLine.mm [hunk 14]
```diff
@@ -16,16 +16,48 @@
...
+    if (result == nullptr) {
+        return nullptr;
+    }
+
+    return CFStringCreateWithBytesNoCopy(kCFAllocatorDefault,
+                                         reinterpret_cast<UInt8*>(characters.release()),
+                                         bufLen,
+                                         kCFStringEncodingUTF16,
+                                         false,
+                                         nullptr);
```
Comment (DHowett-MSFT) [inline Frameworks/CoreText/CTLine.mm]: you will need to provide an allocator to delete this string; look at the implementation of `initWithBytesNoCopy:freeWhenDone:` for NSString, expecting freeWhenDone=YES

File: Frameworks/CoreText/CTLine.mm (hunk 13 above)
Comment (msft-Jeyaram) [inline Frameworks/CoreText/CTLine.mm]: 🤣  that was coming 😄

**Comment by msft-Jeyaram** [comment]
@DHowett-MSFT please take a look

File: Frameworks/CoreText/CTLine.mm (hunk 14 above)
Comment (msft-Jeyaram) [inline Frameworks/CoreText/CTLine.mm]: Already passing in the nullptr for deallocation.
Switched to use characters, which does it for you.

File: Frameworks/CoreText/CTLine.mm [hunk 15]
```diff
@@ -16,16 +16,46 @@
...
+static CFStringRef __CTCreateReversedString(CFStringRef string) {
+    if (string == nullptr) {
+        return nullptr;
+    }
+
+    CFIndex length = CFStringGetLength(string);
+    CFIndex usedBufLen;
+    CFStringGetBytes(string, CFRangeMake(0, length), kCFStringEncodingUTF16, 0, false, nullptr, length, &usedBufLen);
+
+    if (length < 2) {
```
Comment (aballway) [inline Frameworks/CoreText/CTLine.mm]: nit: move this above CFStringGetBytes call

File: Frameworks/include/CFFoundationInternal.h [hunk 16]
```diff
@@ -1,6 +1,6 @@
 //******************************************************************************
 //
-// Copyright (c) 2015 Microsoft Corporation. All rights reserved.
+// Copyright (c) Microsoft. All rights reserved.
```
Comment (aballway) [inline Frameworks/include/CFFoundationInternal.h]: These clean-ups, though nice, seem unrelated to this PR.  Just revert these files?
